etl-pipeline/processes/record_candidate_finder.py

```diff
@@ -0,0 +1,194 @@
+import re
+from logging import Logger
+from typing import List, Set, Tuple, Optional, Any
+from sqlalchemy.exc import DataError
+from managers import DBManager
+
+
+from model import Record
+from logger import create_log
+
+logger = create_log(__name__)
+
+
+class CandidateRecordFinder:
+    
+    # Maximum number of "hops" to follow when matching records through identifiers
+    MAX_MATCH_DISTANCE = 4
+    # Maximum number of records that can be in a candidate pool
+    CANDIDATE_POOL_LIMIT = 10000
+    # Regular expression to identify identifiers that should be matched
+    IDENTIFIERS_TO_MATCH = r"\|(?:isbn|issn|oclc|lccn|owi)$"
+    
+    def __init__(self, db_manager: DBManager):
+        self.db_manager = db_manager
+    
+    def find_candidate_records(self, record: Record) -> List[Record]:
+        """Find records that might be related to the input record.
+        Raises:
+            Exception: If the number of candidate records exceeds CANDIDATE_POOL_LIMIT
```

Comment:
Hmm could we not throw an exception here and just cap the number of records returned?

Reply:
I had that thought too but didn't think I knew enough to change it yet : ). Happy to simply cap it though, especially if I can send a warning to NewRelic when that happens

Reply:
Ya we can make the change after!